From 798a358c72532a063d64f179dde1496bc2cf9400 Mon Sep 17 00:00:00 2001
From: Leo Yu-Chi Liang <ycliang@andestech.com>
Date: Tue, 27 Aug 2024 15:36:08 +0800
Subject: [PATCH 076/168] cache: andes: Improve performance of L2 cache
 operation (#170)

Inspired by [ticket #26566](https://es.andestech.com/scp/tickets.php?id=26566),  we could reduce L2 cache operation time by
eliminating get_cpu() call on !CONFIG_SMP and eliminating the fence instructions on I/O memory access.

The platform specification defines that all I/O regions are on channel 0 (i.e. point-to-point strongly ordered),
so explicit fence instructions are not required. [1] [2] [3]

[1] platform spec 2.1.1: https://github.com/riscvarchive/riscv-platform-specs/blob/main/riscv-platform-spec.adoc?plain=1#L169
[2] privileged spec 3.6.5: https://github.com/riscv/riscv-isa-manual/blob/main/src/machine.adoc?plain=1#L2835
[3] riscv: asm/mmio.h: https://gitea.andestech.com/RD-SW/linux/src/branch/ast-v5_4_0-branch/arch/riscv/include/asm/mmio.h#L105

Signed-off-by: Leo Yu-Chi Liang <ycliang@andestech.com>
Reviewed-on: https://gitea.andestech.com/RD-SW/linux/pulls/170
Reviewed-by: Peter Yu-Chien Lin <peterlin@andestech.com>
Reviewed-by: Charles Ci-Jyun Wu <dminus@andestech.com>
Reviewed-by: Tim Shih-Ting OuYang <tim609@andestech.com>
---
 drivers/cache/andes_cache.c | 23 ++++++++++++++---------
 1 file changed, 14 insertions(+), 9 deletions(-)

diff --git a/drivers/cache/andes_cache.c b/drivers/cache/andes_cache.c
index 037020c13c68..4aa24168198d 100644
--- a/drivers/cache/andes_cache.c
+++ b/drivers/cache/andes_cache.c
@@ -115,27 +115,27 @@ static irqreturn_t l2c_irq(int irq, void *dev_id)
 /* L2 Cache operations */
 static inline uint32_t andes_cpu_l2c_get_cctl_status(void)
 {
-	return readl(andes_priv.l2c_base + ANDES_L2C_REG_STATUS_OFFSET);
+	return readl_relaxed(andes_priv.l2c_base + ANDES_L2C_REG_STATUS_OFFSET);
 }
 
 static void cpu_l2c_cctl(phys_addr_t pa, void __iomem *base,
 			 int mhartid, unsigned long ops)
 {
 #ifdef CONFIG_64BIT
-	writeq(pa, (base + ANDES_L2C_REG_CN_ACC_OFFSET(mhartid)));
+	writeq_relaxed(pa, (base + ANDES_L2C_REG_CN_ACC_OFFSET(mhartid)));
 #else
 	/*
 	 * Considering RV32 potential to use over 4G memory,
 	 * the physical address is split into upper and lower 32 bits
 	 * and stored in a 64-bits PA-type L2C CCTL access line register.
 	 */
-	writel((pa & 0xFFFFFFFF),
-	       (base + ANDES_L2C_REG_CN_ACC_OFFSET(mhartid)));
-	writel((pa >> 32),
-	       (base + ANDES_L2C_REG_CN_ACC_OFFSET(mhartid) + 0x4));
+	writel_relaxed((pa & 0xFFFFFFFF),
+		       (base + ANDES_L2C_REG_CN_ACC_OFFSET(mhartid)));
+	writel_relaxed((pa >> 32),
+		       (base + ANDES_L2C_REG_CN_ACC_OFFSET(mhartid) + 0x4));
 #endif /* !CONFIG_64BIT */
 
-	writel(ops, base + ANDES_L2C_REG_CN_CMD_OFFSET(mhartid));
+	writel_relaxed(ops, base + ANDES_L2C_REG_CN_CMD_OFFSET(mhartid));
 	while ((andes_cpu_l2c_get_cctl_status() &
 		ANDES_CCTL_L2_STATUS_CN_MASK(mhartid)) !=
 		ANDES_CCTL_L2_STATUS_IDLE)
@@ -147,8 +147,11 @@ static void andes_cpu_cache_operation(unsigned long start, unsigned long end,
 {
 	unsigned long line_size = andes_priv.andes_cache_line_size;
 	void __iomem *base = andes_priv.l2c_base;
-	int mhartid = get_cpu();
 	unsigned long pa;
+	int mhartid = 0;
+
+	if (IS_ENABLED(CONFIG_SMP))
+		mhartid = get_cpu();
 
 	if (likely(base)) {
 		while (end > start) {
@@ -168,7 +171,9 @@ static void andes_cpu_cache_operation(unsigned long start, unsigned long end,
 			start += line_size;
 		}
 	}
-	put_cpu();
+
+	if (IS_ENABLED(CONFIG_SMP))
+		put_cpu();
 }
 
 /* Write-back L1 and L2 cache entry */
-- 
2.34.1

